pandas/io/sql.py

```diff
@@ -189,6 +190,18 @@ def execute(sql, con, params=None):
 # -- Read and write to DataFrames
 
 
+def _read_helper(method: Callable, iterate: bool, **kwargs):
+    """method returns an iterator.  If iterate is False, we need to exhaust the iterator
```

Comment:
Can there be a more direct fix to leave the connection open if chunksize is provided? It seems fairly roundabout to exhaust the iterator?

Reply:
To keep the connection open, I need to make sure that the context manager at line 864 stays open.  My solution is to create generating functions `_read_sql_table`, `_read_sql_query`, and `_read_sql`.  If `chunksize=None`, these generating functions return a single `DataFrame`.  Then I have the opposite problem: if the caller wants a single DataFrame, I have to exit out of the context manager at line 864 so that the connection will close.  If instead of
```
for frame in result:
    pass
return frame
```
I use
```
return next(result)
```
then the line 864 context manager would not exit.  I agree it is roundabout, but I think having `SQLDatabase` only accept a `Connection`, not an `Engine`, and having `SQLDatabase.read_query` only return an `Iterator[DataFrame]` also simplifies things a bit.

Reply:
Can `_sqlalchemy_con` be aware of when `chunksize` is specified and keep the connection open as needed?

Reply:
We are thinking along similar lines.  I think it's will be easier to fold `_sqlalchemy_con` into the `SQLDatabase` class, because I want the connection to be closed at the end of iteration, or if there is an exception.